aten/src/ATen/native/cpu/moments_utils.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <array>
+#include <cstring>
+#include <utility>
+#include <vector>
+
+#include <ATen/cpu/vec256/vec256.h>
+#include <ATen/native/cpu/utils.h>
+#include <c10/util/SmallVector.h>
+
+namespace at {
+namespace native {
+namespace utils {
+
+template <typename T>
+void AddMoments(
+    int64_t m0_add,
+    const T& m1_add,
+    const T& m2_add,
+    int64_t& m0,
+    T& m1,
+    T& m2) {
+  const int64_t n = m0 + m0_add;
+  const T c1 = n == 0 ? 0 : static_cast<T>(m0) / static_cast<T>(n);
+  const T c2 = n == 0 ? 0 : T(1) - c1;
+  const T delta = m1_add - m1;
+  m0 = n;
+  m1 = c1 * m1 + c2 * m1_add;
+  m2 += m2_add + delta * delta * c1 * static_cast<T>(m0_add);
+}
+
+template <typename T>
+void AddMomentsVec(
+    int64_t m0_add,
+    const vec256::Vec256<T>& m1_add,
+    const vec256::Vec256<T>& m2_add,
+    int64_t& m0,
+    vec256::Vec256<T>& m1,
+    vec256::Vec256<T>& m2) {
+  using Vec = vec256::Vec256<T>;
+  const int64_t n = m0 + m0_add;
+  const T c1 = n == 0 ? 0 : static_cast<T>(m0) / static_cast<T>(n);
+  const T c2 = n == 0 ? 0 : T(1) - c1;
+  const Vec c1_vec(c1);
+  const Vec c2_vec(c2);
+  const Vec delta = m1_add - m1;
+  m0 = n;
+  m1 = c1_vec * m1 + c2_vec * m1_add;
+  m2 += m2_add + delta * delta * c1_vec * Vec(static_cast<T>(m0_add));
+}
+
+// Compute rowwise moments by Welford algorithm and cascade sum to improve
+// numerical stability.
+// https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+// https://en.wikipedia.org/wiki/Pairwise_summation
+template <typename T>
+std::pair<T, T> RowwiseMoments(const T* X, int64_t N) {
+  using Vec = vec256::Vec256<T>;
+
+  constexpr int64_t kVecSize = Vec::size();
+  constexpr int64_t kChunkSize = 32;
+  constexpr int64_t kMaxDepth = 64;
```

Comment:
max depth of 64 seems like an overkill (given that it is then used as the size for the m0_stk, m1_stk, m2_stk variables)? In the reduction PR #39516 there were just 4 levels and that seemed enough for pretty large tensors.

Reply:
Personally I think it will be more reliable to keep a fixed chunk size, especially for the case in var because there is square op for the input. I changed the chunk size to 128 which is the same as numpy below. And also make the kMaxDepth as a parameter for the template function. Then if the size is small, the SmallVector will be smaller to improve the register efficiency.
https://github.com/numpy/numpy/blob/623bc1fae1d47df24e7f1e29321d0c0ba2771ce0/numpy/core/src/umath/loops_utils.h.src#L60

Reply:
I chose a fixed number of levels because I expect the compiler to do a lot better at allocating registers for a fixed number of accumulators. I don't think a dynamically-sized array of vector is even possible to keep in registers.

For four levels with a chunk size of 16, and a vector size of 8 (for float), you can sum up to 2^19 or .5 million elements. With 5 levels and a chunk size of 16, you can sum 2^63 elements i.e. more floats than you can possibly address with 64-bits.

With 64 levels and a chunk size of 16, you could count every atom in the universe.